### PR TITLE
Add connection pooling

### PR DESCRIPTION
**Pull request by kevinnoel-be**
**Description:** <Describe what has changed.>
Reuse of connections created per database (configured or discovered) vs current behavior to create & close connection per database on each scrape.

**Link to tracking Issue:** 
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30831

**Testing:** 
Updated unit & integration tests. Also, ran locally multiple scenario:
- no feature gate specified (default): current behavior maintained, connections created/closed on each database per scrape
- feature gate connection pool enabled, no connection pool config specified (default): reduction of the number of connections created/closed
- feature gate connection pool enabled, connection pool config tweaked: connections created on first scrape & closed when configured lifetime reached or collector shutdown

**Documentation:**
- change log
- readme for the feature gate & related optional configurations linked to this feature

**Note**
Checking internally for getting the CLA signed

### CODE REVIEW

File: receiver/postgresqlreceiver/client.go [hunk 1]
```diff
@@ -102,41 +101,32 @@ func sslConnectionString(tls configtls.TLSClientSetting) string {
 	return conn
 }
 
-func newPostgreSQLClient(conf postgreSQLConfig) (*postgreSQLClient, error) {
+func (c postgreSQLConfig) DSN() (string, error) {
```
Comment (djaglowski) [inline receiver/postgresqlreceiver/client.go:104]: DSN? Can we use a more expressive name?

**Review by djaglowski — COMMENTED** [review]
LGTM, just one nit

File: receiver/postgresqlreceiver/client.go (hunk 1 above)
Comment (kevinnoel-be) [reply receiver/postgresqlreceiver/client.go:104]: Sure, it's a bit cryptic indeed. Maybe `ConnectionString`?

Comment (djaglowski) [reply receiver/postgresqlreceiver/client.go:104]: Sure, sounds good to me.

Comment (kevinnoel-be) [reply receiver/postgresqlreceiver/client.go]: Done

**Comment by djaglowski** [comment]
@kevinnoel-be, please sign the CLA

**Comment by kevinnoel-be** [comment]
> @kevinnoel-be, please sign the CLA

As mentioned in the description, working on it internally to get it signed, please bear with us 😅

**Comment by kevinnoel-be** [comment]
@djaglowski Thanks for you patience, the CLA is now signed!

**Comment by djaglowski** [comment]
Thanks @kevinnoel-be. Code looks generally good to me but I think some recent changes upstream have broken compilation: `client_factory.go:42:18: cfg.NetAddr undefined`

**Comment by kevinnoel-be** [comment]
@djaglowski My bad, didn't check much on my side. Ran tests + integration tests locally before (properly) pushing again this time 😅

**Comment by djaglowski** [comment]
> My bad, didn't check much on my side. 

No worries. There's a decent gap between local tooling and CI, and lots of unrelated activity causing minor conflicts, so it can be very difficult.

Looks like one more check: `Generated code is out of date, please run "make generate" and commit the changes in this PR.`

**Comment by kevinnoel-be** [comment]
> > My bad, didn't check much on my side.
> 
> No worries. There's a decent gap between local tooling and CI, and lots of unrelated activity causing minor conflicts, so it can be very difficult.
> 
> Looks like one more check: `Generated code is out of date, please run "make generate" and commit the changes in this PR.`

Uh, I guess there's an order to things. I did run `gci` which changed those imports, maybe needs to be fixed in some template to avoid this? I'll take a look

**Comment by djaglowski** [comment]
Hmm, I think this may be related to recent changes in out generated test code. I ran `make generate` locally on this branch and get the same incorrect format. cc: @atoulme 

```
@@ -9,10 +9,12 @@ import (
 	"github.com/stretchr/testify/require"
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/component/componenttest"
-	"go.opentelemetry.io/collector/confmap/confmaptest"
+
 	"go.opentelemetry.io/collector/consumer/consumertest"
 	"go.opentelemetry.io/collector/receiver"
 	"go.opentelemetry.io/collector/receiver/receivertest"
+
+	"go.opentelemetry.io/collector/confmap/confmaptest"
 )
 
 // assertNoErrorHost implements a component.Host that asserts that there were no errors.
Generated code is out of date, please run "make generate" and commit the changes in this PR.
```

[link to failure here](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8097348667/job/22130705504?pr=31127#step:16:301)

**Comment by atoulme** [comment]
You installed mdatagen from core - try `make generate` in contrib. I'll harmonize the differences between the repos asap.

**Comment by djaglowski** [comment]
Thanks @atoulme. Unfortunately I ran `make generate` in contrib (and did again just now with the same result).

```
➜  opentelemetry-collector-contrib (config-db-pool) make generate
cd cmd/mdatagen && go install .
/Applications/Xcode.app/Contents/Developer/usr/bin/make for-all CMD="go generate ./..."
running go generate ./... in root
running go generate ./... in ./cmd/configschema
...
running go generate ./... in ./receiver/postgresqlreceiver
...
running go generate ./... in ./testbed
running go generate ./... in ./testbed/mockdatasenders/mockdatadogagentexporter
➜  opentelemetry-collector-contrib (config-db-pool) git st                                                                                                                               ✭
On branch config-db-pool
Your branch and 'collibra/config-db-pool' have diverged,
and have 24 and 1 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   receiver/postgresqlreceiver/generated_component_test.go

no changes added to commit (use "git add" and/or "git commit -a")
➜  opentelemetry-collector-contrib (config-db-pool) git diff                                                                                                                             ✭
diff --git a/receiver/postgresqlreceiver/generated_component_test.go b/receiver/postgresqlreceiver/generated_component_test.go
index 42ffa42608..9b09d44ee2 100644
--- a/receiver/postgresqlreceiver/generated_component_test.go
+++ b/receiver/postgresqlreceiver/generated_component_test.go
@@ -9,10 +9,12 @@ import (
        "github.com/stretchr/testify/require"
        "go.opentelemetry.io/collector/component"
        "go.opentelemetry.io/collector/component/componenttest"
-       "go.opentelemetry.io/collector/confmap/confmaptest"
+
        "go.opentelemetry.io/collector/consumer/consumertest"
        "go.opentelemetry.io/collector/receiver"
        "go.opentelemetry.io/collector/receiver/receivertest"
+
+       "go.opentelemetry.io/collector/confmap/confmaptest"
 )
```

**Comment by atoulme** [comment]
You might want to rebase off the latest main, at least with https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31520 the generated files are impacted.

**Comment by kevinnoel-be** [comment]
AFAIS the order we see in CI and locally when running `generate` seems to correspond with what is in the [template](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/cef707edcfd7d943a1bc74815aa3d9be07be92e8/cmd/mdatagen/templates/component_test.go.tmpl#L28-L41). I had the "good" idea to run [`gci` target](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/cef707edcfd7d943a1bc74815aa3d9be07be92e8/Makefile.Common#L224) at some point _after_ the `generate` one which is why this check is failing (as the template doesn't seem to have the "proper" import order and `gci` didn't ignore the generated file).

TL;DR I'll rebase on Monday after a `make generate` to fix this PR. To avoid any similar issue I guess the generated template should follow the imposed/configured imports order.

**Review by djaglowski — APPROVED** [review]
